frontend/website/src/components/chat/chatroom.svelte

```diff
@@ -0,0 +1,443 @@
+
+	let collapsed: boolean = browser && window.localStorage.getItem("chatroom_collapsed") === "true";
+	$: {
+		if (browser) {
+			window.localStorage.setItem("chatroom_collapsed", JSON.stringify(collapsed));
+		}
+	}
+
+	export let channelId: string;
+
+	function collapseNav() {
+		collapsed = !collapsed;
+	}
+
+	let messagesEl: HTMLDivElement;
+	let chatStatus = writable(ChatStatus.Connecting);
+	let chatMessage = "";
+	let atBottom = true;
+
+	$: inputEmpty = chatMessage.length === 0 || chatMessage === "\n";
+
+	$: inputValid = chatMessage.length <= MAX_MESSAGE_LENGTH;
+
+	$: sendEnabled = inputValid && !inputEmpty && $chatStatus === ChatStatus.Connected && $user;
+
+	$: chatError =
+		$user === null
+			? "Please create an account to chat"
+			: chatMessage.length > MAX_MESSAGE_LENGTH
+			? `Your message is longer than ${MAX_MESSAGE_LENGTH} characters`
+			: null;
+
+	function onScroll() {
+		atBottom = messagesEl.scrollTop + messagesEl.offsetHeight >= messagesEl.scrollHeight - 30;
+	}
+
+	onMount(() => {
+		// Scroll to bottom
+		messagesEl.addEventListener("scroll", onScroll);
+		return () => {
+			if (messagesEl) {
+				messagesEl.removeEventListener("scroll", onScroll);
+			}
+		};
+	});
+
+	function scrollToBottom() {
+		// We set at bottom to true so that if new messages are added we scroll to the bottom again
+		atBottom = true;
+		// We request an animation frame to make sure the scroll happens after the DOM has been updated
+		window.requestAnimationFrame(() => {
+			if (messagesEl) {
+				messagesEl.scrollTop = messagesEl.scrollHeight;
+			}
+		});
+	}
+
+	afterUpdate(() => {
+		if (atBottom) {
+			scrollToBottom();
+		}
+	});
+
+	const newMessageQuery = graphql(`
+		mutation SendMessage($channelId: UUID!, $content: String!) {
+			chat {
+				sendMessage(channelId: $channelId, content: $content) {
+					id
+				}
+			}
+		}
+	`);
+
+	async function sendMessageInner(message: string) {
+		const response = await client
+			.mutation(newMessageQuery, { channelId: channelId, content: message })
+			.toPromise();
+		if (response.error) {
+			// TODO: Failed to send message
+		}
+	}
+
+	function onChatmessageKeydown(e: KeyboardEvent) {
+		if (e.key === "Enter" && !e.shiftKey) {
+			e.preventDefault();
+			sendMessage();
+		}
+		e.stopPropagation();
+	}
+
+	function onChatmessagePaste(e: ClipboardEvent) {
+		// To prevent inserting rich text
+		if (e.clipboardData) {
+			const text = e.clipboardData.getData("text/plain");
+			// This is deprecated but it has no good alternative.
+			// https://developer.mozilla.org/en-US/docs/Web/API/Document/execCommand
+			// https://stackoverflow.com/a/70831583/10772729
+			// TL;DR: This will most likely be never removed from any major browser
+			document.execCommand("insertText", false, text);
+			// When execCommand is removed or is unsupported,
+			// it will fall back to the default rich text paste, because an error will occur before preventDefault is called.
+			e.preventDefault();
```

Comment:
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/contenteditable

Does this work `contenteditable="plaintext-only"`

Reply:
I didn't see this when I implemented it. Good idea. Will probably work.

Reply:
It isn't widely supported yet. But I'll add a comment.
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/contenteditable#browser_compatibility
https://caniuse.com/mdn-html_global_attributes_contenteditable_plaintext-only

Reply:
Ah i see.